x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/index/mapper/ConstantKeywordFieldMapper.java

```diff
@@ -0,0 +1,243 @@
+        }
+
+        @Override
+        public String typeName() {
+            return CONTENT_TYPE;
+        }
+
+        @Override
+        public IndexFieldData.Builder fielddataBuilder(String fullyQualifiedIndexName) {
+            return new ConstantKeywordIndexFieldData.Builder(mapperService -> value);
+        }
+
+        private static String valueToString(Object v) {
+            if (v instanceof BytesRef) {
+                return ((BytesRef) v).utf8ToString();
+            } else {
+                return v.toString();
+            }
+        }
+
+        @Override
+        public Query termQuery(Object value, QueryShardContext context) {
+            if (Objects.equals(valueToString(value), this.value)) {
+                return new MatchAllDocsQuery();
+            } else {
+                return new MatchNoDocsQuery();
+            }
+        }
+
+        @Override
+        public Query termsQuery(List<?> values, QueryShardContext context) {
+            for (Object v : values) {
+                if (Objects.equals(valueToString(v), value)) {
+                    return new MatchAllDocsQuery();
+                }
+            }
+            return new MatchNoDocsQuery();
+        }
+
+        @Override
+        public Query prefixQuery(String value,
+                @Nullable MultiTermQuery.RewriteMethod method,
+                QueryShardContext context) {
+            if (this.value.startsWith(value)) {
+                return new MatchAllDocsQuery();
+            } else {
+                return new MatchNoDocsQuery();
+            }
+        }
+
+        public Query wildcardQuery(String value,
+                @Nullable MultiTermQuery.RewriteMethod method,
+                QueryShardContext context) {
+            if (Regex.simpleMatch(value, this.value)) {
+                return new MatchAllDocsQuery();
+            } else {
+                return new MatchNoDocsQuery();
+            }
+        }
+    }
+
+    protected ConstantKeywordFieldMapper(String simpleName, MappedFieldType fieldType, MappedFieldType defaultFieldType,
+                                 Settings indexSettings) {
+        super(simpleName, fieldType, defaultFieldType, indexSettings, MultiFields.empty(), CopyTo.empty());
+    }
+
+    @Override
+    protected ConstantKeywordFieldMapper clone() {
+        return (ConstantKeywordFieldMapper) super.clone();
+    }
+
+    @Override
+    public ConstantKeywordFieldType fieldType() {
+        return (ConstantKeywordFieldType) super.fieldType();
+    }
+
+    @Override
+    protected void parseCreateField(ParseContext context, List<IndexableField> fields) throws IOException {
+        String value;
+        if (context.externalValueSet()) {
+            value = context.externalValue().toString();
+        } else {
+            XContentParser parser = context.parser();
+            if (parser.currentToken() == XContentParser.Token.VALUE_NULL) {
+                value = fieldType().nullValueAsString();
```

Comment:
I don't think we'll ever have a `null_value` set, we could just skip this check?